### PR TITLE
Add IPv4 variant to Host and remove custom Ipv6 address type

### PR DESCRIPTION
**Pull request by pyfisch**
The Whatwg URL Standard was changed to describe Ipv4 address
parsing. Before this change there was no difference made
between domains and Ipv4 addresses. This change also removes
the custom Ipv6 type and uses the types provided by std for
ip addresses.

https://url.spec.whatwg.org/#syntax-host-ipv4

Note `heap_size` feature is broken because the HeapSizeOf trait is not yet implemented for the ip addresses.

[Review on Reviewable](https://reviewable.io/reviews/servo/rust-url/135)

### CODE REVIEW

**Comment by pyfisch** [comment]
https://github.com/pyfisch/heapsize/tree/4bdd4e2524e506d9dd23115dff6a9a2249fe6cc4 implements `HeapSizeOf` for ip addresses.

**Comment by pyfisch** [comment]
By the way this solves https://github.com/servo/rust-url/issues/116

**Comment by untitaker** [comment]
@pyfisch I think you should file a PR against heapsize regarding that commit.

**Comment by Manishearth** [comment]
Blocked on https://github.com/servo/heapsize/pull/17 getting r+ since crates and the repo weren't in sync.

**Comment by Ms2ger** [comment]
Please bump the heapsize dep to `0.1.3`.

**Comment by pyfisch** [comment]
done

**Comment by Ms2ger** [comment]
And squash the commits, please

**Comment by pyfisch** [comment]
done

**Comment by Ms2ger** [comment]
Thanks!

File: src/host.rs [hunk 1]
```diff
@@ -51,22 +37,30 @@ impl Host {
...
-            let decoded = percent_decode(input.as_bytes());
-            let domain = String::from_utf8_lossy(&decoded);
-            // TODO: Remove this check and use IDNA "domain to ASCII"
-            if !domain.is_ascii() {
-                Err(ParseError::NonAsciiDomainsNotSupportedYet)
-            } else if domain.find(&[
-                '\0', '\t', '\n', '\r', ' ', '#', '%', '/', ':', '?', '@', '[', '\\', ']'
-            ][..]).is_some() {
-                Err(ParseError::InvalidDomainCharacter)
+            if let Ok(addr) = input.parse() {
```
Comment (SimonSapin) [inline src/host.rs]: @pyfisch @Ms2ger have you verified that `Ipv4Addr::parse` and `Ipv6::parse` match https://url.spec.whatwg.org/#host-parsing in every edge case? For example: leading zeros, hex/octal components in v4, …

**Comment by SimonSapin** [comment]
I’ve reverted this change. It looks like there was a misunderstanding in the review, sorry. At least the IPv4 parsing of the standard library differs from https://url.spec.whatwg.org/#concept-ipv4-parser . (Note that the "parsing" and "serialization" sections of the spec are more relevant to implementation than the "syntax" sections, which only give a high level overview as opposed to exact algorithms.)

I don’t know if the current parsing or serialization behavior of std for *IPv6* happens to match the current spec, but I don’t think we should rely on it staying that way. (I don’t know if it should. Web compatibility with all its weirdness is an overriding concern for browsers, but not as much for a programming language standard library.)

@pyfisch, could you please open a new PR? But keep the existing `fmt` and `parse` methods for IPv6, and write IPv4 parsing per https://url.spec.whatwg.org/#concept-ipv4-parser . Or let us know if you’d rather someone else does this work.

**Comment by SimonSapin** [comment]
(Edge cases to look for in IPv4 include: leading zeroes, hexadecimal or octal components, less than 4 components.)

**Comment by pyfisch** [comment]
I will create a new pull request which will use custom parse methods as they are described in the urlspec. The canonical formatting is unambiguous so I will try to use the methods provided by std but add many tests for edge cases.

**Comment by SimonSapin** [comment]
Please also add "Fixes #116" in the new PR or commit message. Thanks!
